numpyro/distributions/multivariate.py

```diff
@@ -7,27 +7,80 @@
 # All rights reserved.
 
 import jax.numpy as np
+from jax import lax
+from jax.experimental.stax import softmax
+from jax.numpy.lax_numpy import _promote_dtypes
 from jax.scipy.special import digamma, gammaln
 
 from numpyro.distributions import constraints
-from numpyro.distributions.distribution import jax_mvcontinuous
-from numpyro.distributions.util import standard_gamma, xlogy
+from numpyro.distributions.discrete import binom
+from numpyro.distributions.distribution import jax_continuous, jax_discrete, jax_multivariate
+from numpyro.distributions.util import categorical_rvs, entr, multinomial_rvs, standard_gamma, xlogy
 
 
 def _lnB(alpha):
     return np.sum(gammaln(alpha), axis=-1) - gammaln(np.sum(alpha, axis=-1))
 
 
-class dirichlet_gen(jax_mvcontinuous):
-    # TODO: use dirichlet doc instead of the default one of rv_continuous
+# TODO: either use multivariate docs instead of the default one of
+# rv_continuous/rv_discrete
+
+class categorical_gen(jax_multivariate, jax_discrete):
+    @property
+    def arg_constraints(self):
+        if self.is_logits:
+            return {'p': constraints.real}
+        else:
+            return {'p': constraints.simplex}
+
+    def _support(self, *args, **kwargs):
+        (p,), _, _ = self._parse_args(*args, **kwargs)
+        return constraints.integer_interval(0, p.shape[-1] - 1)
+
+    def _batch_shape(self, p):
+        return p.shape[:-1]
+
+    def _event_shape(self, p):
+        return ()
```

Comment:
`event_shape` is not required but it is useful for testing, so I add it just for convenience.